src/parsing/chunked_reader.rs

```diff
@@ -34,95 +33,65 @@ where
+            reached_eof: false,
         }
     }
 
-    #[inline]
-    fn remaining(&self) -> u64 {
-        self.length - self.read
-    }
-
-    #[inline]
-    fn read_line(&mut self) -> io::Result<usize> {
-        buffers::read_line(&mut self.inner, &mut self.line, 128)
-    }
-
-    fn read_chunk_size(&mut self) -> io::Result<u64> {
-        self.read_line()?;
-        if self.line.is_empty() {
+    fn read_chunk_size(&mut self) -> io::Result<usize> {
+        buffers::read_line(&mut self.inner, &mut self.buffer, 128)?;
+        if self.buffer.is_empty() {
             return Err(io::ErrorKind::UnexpectedEof.into());
         }
-        parse_chunk_size(&self.line)
-    }
-
-    fn read_empty_line(&mut self) -> io::Result<()> {
-        let n = self.read_line()?;
-        if n == 0 || !self.line.is_empty() {
-            Err(InvalidResponseKind::Chunk.into())
-        } else {
-            Ok(())
-        }
+        parse_chunk_size(&self.buffer)
     }
 }
 
-impl<R> Read for ChunkedReader<R>
+impl<R> BufRead for ChunkedReader<R>
 where
     R: Read,
 {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        if self.is_expecting_chunk {
-            debug!("waiting, parsing new chunk size");
-            // If we're waiting for a new chunk, we read a line and parse the number as hexadecimal.
-            self.read = 0;
-            self.length = self.read_chunk_size()?;
-            // If the chunk's length is 0, we've received the EOF chunk.
-            if self.length == 0 {
-                debug!("received EOF chunk");
-                // Read CRLF
-                self.read_empty_line()?;
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        const MAX_BUFFER_LEN: usize = 64 * 1024;
+
+        if self.buffer.len() == self.consumed && !(self.remaining == 0 && self.reached_eof) {
+            if self.remaining == 0 {
+                self.remaining = self.read_chunk_size()?;
+                if self.remaining == 0 {
+                    self.reached_eof = true;
+                }
             }
-            self.is_expecting_chunk = false;
-        }
-
-        // If we have a length of 0 for a chunk, we've reached EOF.
-        if self.length == 0 {
-            return Ok(0);
-        }
-
-        // We read the smallest amount between the given buffer's length and the remaining bytes' length.
-        let count = cmp::min(self.remaining(), buf.len() as u64) as usize;
 
-        debug!(
-            "before read, remaining={}, count={}, buflen={}",
-            self.remaining(),
-            count,
-            buf.len()
-        );
+            self.buffer.resize(cmp::min(self.remaining, MAX_BUFFER_LEN), 0);
+            self.inner.read_exact(&mut self.buffer)?;
+            self.consumed = 0;
+            self.remaining -= self.buffer.len();
 
-        let n = self.inner.read(&mut buf[..count])?;
-        self.read += n as u64;
+            if self.remaining == 0 && !buffers::read_line_ending(&mut self.inner)? {
```

Comment:
I'm not sure I understand this if, in which will `read_line_ending` return false?

Reply:
Ah if we've read the whole chunk and the next bytes aren't `\r` or `\r\n` it's an error, got it.